### ISSUE TITLE
Non-deterministic failures (exit code 127)

### ISSUE THREAD

**Issue by omh1280** [hadolint/hadolint#475]
- [x] This is a bug report
- [ ] This is a feature request
- [x] I searched existing issues before opening this one

### Expected behavior

Hadolint will exit with code `0` if run against a valid Dockerfile.

### Actual behavior

Hadolint rarely exits with code `127` on a Dockerfile that should pass.

This is reproducible on Mac and Linux and can be tested with the below dockerfile. Since this bug seems non-deterministic, the value of `1000` is chosen to sufficiently increase the probability that the run will fail.

We use `parallel` to speed up running hadolint 1000 times. This has been seen on CI/CD infrequently, which is why this issue is being raised.

### Steps to reproduce the behavior

**Bash**
```bash
$ mkdir hadolint-stress && cd hadolint-stress
$ touch Dockerfile   # copy dockerfile below into here
$ docker build -t hadolint-stress .
$ docker run hadolint-stress   # should exit with 127 most of the time
```

**Dockerfile**

```dockerfile
FROM ubuntu:focal

RUN apt-get update && apt-get install --no-install-recommends -y \
  wget=1.20.3-1ubuntu1 \
  ca-certificates=20190110ubuntu1.1 \
  parallel=20161222-1.1 \
  && rm -rf /var/lib/apt/lists
RUN wget -q https://github.com/hadolint/hadolint/releases/download/v1.18.0/hadolint-Linux-x86_64 -O hadolint \
  && chmod +x hadolint

COPY ./Dockerfile ./Dockerfile

# hadolint ignore=SC2039
CMD [ "/bin/bash", "-c", "parallel -N0 --halt now,fail=1 ./hadolint ./Dockerfile ::: {1..1000}; echo $?" ]
```

If the value of `1000` is reduced to `100`, a majority of the runs pass:

```bash
$ docker run hadolint-stress
0
$ docker run hadolint-stress
0
$ docker run hadolint-stress
0
$ docker run hadolint-stress
0
$ docker run hadolint-stress
0
$ docker run hadolint-stress
parallel: This job failed:
./hadolint ./Dockerfile
127
```

**Output of `hadolint --version` or
  `docker run --rm hadolint/hadolint hadolint --version`:**  Haskell Dockerfile Linter v1.18.0-0-g76eee5c (seen in 1.17 as well)

**Additional environment details (OS, stack version, etc.)** MacOS & Linux

**Comment by m-ildefons** [comment]
Hi @omh1280 ,
I am not sure that this is an issue with Hadolint. Exit code 127 means: "Executable not found in $PATH".
While I could reproduce the issue inside a docker container, I fail to reproduce the issue on a 'bare metal' host, although my setup there is significantly different from the setup with the docker file:

- OS: Debian buster
- Hadolint: Haskell Dockerfile Linter v1.18.0-24-gb40ea64
- `parallel`: from `moreutils` 0.62-1
- `bash`:  5.0.3(1)-release

I could run this without trouble:
```
> for i in {1..100}; do parallel sh -c "hadolint Dockerfile" -- {1..1000}; done; echo $?
0
```

You can also try and use bash's builtin job features:
```
> for i in {1..1000}; do hadolint Dockerfile & done
```

**Comment by omh1280** [comment]
Hi @m-ildefons, thanks for investigating!

This is reproducible on my bare-metal Linux box running:

* Ubuntu 20.04
* Bash 5.0.17
* parallel `20161222`

```bash
$ for i in {1..1000}; do hadolint Dockerfile & done
[627]   Exit 127                hadolint Dockerfile
```

Sometimes this and the Dockerfile pass, though, so I usually bump 1000 up to an even higher number. Changing the base image to `debian:buster` also does not fix this issue. I'll continue investigating this weekend.

I've only found [one potentially relevant link](https://www.jamf.com/jamf-nation/discussions/17322/problems-with-good-scripts-failing-exit-code-127) so far on the internet, but this could be unrelated.

**Comment by lorenzo** [comment]
Anecdotally, I have experienced this issue with other binaries in the CI system at work. I wonder if it has anything to do with docker.

That said, I think this issue presents a good case for adding parallelism support to Hadolint. I'll work on that

**Comment by omh1280** [comment]
Reproducible on linux (not in docker) and not in a shell with the following Go code:

```go
package main

import (
        "fmt"
        "log"
        "os/exec"
)

type result struct {
        error
        out []byte
}

func worker(jobs <-chan struct{}, results chan<- result) {
        for range jobs {
                cmd := exec.Command("hadolint", "-v")
                err := cmd.Run()
                if err != nil {
                        results <- result{
                                error: err,
                                out:   nil,
                        }
                }
        }
}

func main() {
        jobs := make(chan struct{})
        results := make(chan result)
        for i := 0; i < 20; i++ {
                go worker(jobs, results)
        }

        for i := 0; i < 5000; i++ {
                select {
                case r := <-results:
                        close(jobs)
                        fmt.Printf("combined out:\n%s\n", string(r.out))
                        log.Fatalf("cmd.Run() failed with %s\n", r.error)
                case jobs <- struct{}{}:
                }
        }

}
```

Note this is just running `hadolint -v`, basically the simplest hadolint command.

```bash
$ go run main.go
```

**Comment by omh1280** [comment]
The reversion occurs between v1.17.2 and v1.17.3

**Comment by lorenzo** [comment]
@omh1280 you mean it does not happen with the newer versions?

**Comment by lorenzo** [comment]
@omh1280 I opened a PR to add parallel linting to hadolint, I'm curious if you would find that feature useful

**Comment by omh1280** [comment]
@lorenzo The bug occurs in all versions since v1.17.3. It was introduced in some commit between v1.17.2 and v1.17.3

**Comment by tanishq-dubey** [comment]
To add a few more data points: 

I was able to replicate these results on _serial_ runs. A simple python test harness can replicate this: 


```python3
import subprocess
import sys

run = 1

while True:
    try:
        subprocess.run("./hadolint-Linux-x86_64 Dockerfile", check=True, shell=True)
        run = run + 1
    except subprocess.CalledProcessError as e:
        print(run, e)
        run = 1
```

Using this script, I was able to get a histogram of how many runs it takes before a crash happens, to see if there was any correlation (could indicated leaked corrupt memory, or some sort of state being held/mutated between runs), however it does not appear to be that way:

![image](https://user-images.githubusercontent.com/3333201/96457974-65a2f880-11ee-11eb-899f-5bf830b475db.png)

In addition to this, I tried to pin down where the crash might be occurring, so I ran strace on two types of runs:

```
for i in {1..5000}; do strace -f bash -c "./hadolint-Linux-x86_64 Dockerfile" >> bashrun 2>&1; done
```
and
```
for i in {1..5000}; do strace -f ./hadolint-Linux-x86_64 Dockerfile >> simplerun 2>&1; done
```

This allowed me to test if `bash` was playing a role in the crash, and from the results it does not seem to be that way. Specifically, in all cases where Hadolint exits with code `127`, the `strace` appears like this:

```
execve("./hadolint-Linux-x86_64", ["./hadolint-Linux-x86_64", "Dockerfile"], 0x7ffeacf7e3b0 /* 89 vars */) = 0
open("/proc/self/exe", O_RDONLY)        = 3
mmap(NULL, 3230335, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f44ffe63000
mmap(0x7f44ffe63000, 3229958, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 3, 0) = 0x7f44ffe63000
mprotect(0x7f4500176000, 6783, PROT_READ|PROT_EXEC) = 0
readlink("/proc/self/exe", "/home/dubey/projects/hadolint-st"..., 4095) = 58
mmap(0x400000, 21479424, PROT_NONE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x400000
mmap(0x400000, 19647156, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x400000
exit(127)                               = ?
+++ exited with 127 +++
```

All good cases appear like this:

```
execve("./hadolint-Linux-x86_64", ["./hadolint-Linux-x86_64", "Dockerfile"], 0x7ffcb80eaf20 /* 89 vars */) = 0
open("/proc/self/exe", O_RDONLY)        = 3
mmap(NULL, 3230335, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fb9707a8000
mmap(0x7fb9707a8000, 3229958, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 3, 0) = 0x7fb9707a8000
mprotect(0x7fb970abb000, 6783, PROT_READ|PROT_EXEC) = 0
readlink("/proc/self/exe", "/home/dubey/projects/hadolint-st"..., 4095) = 58
mmap(0x400000, 21479424, PROT_NONE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x400000
mmap(0x400000, 19647156, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x400000
mprotect(0x400000, 19647156, PROT_READ|PROT_EXEC) = 0
mmap(0x16be000, 1781288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0x12bd000) = 0x16be000
mprotect(0x16be000, 1781288, PROT_READ|PROT_WRITE) = 0
mmap(0x1871000, 42848, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x1871000
mmap(NULL, 4096, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f110098a000
close(3)                                = 0
...
exit(0)                               = ?
+++ exited with 0 +++
```

This occurs in the same way, regardless of Hadolint being invoked with `bash -c` or directly

> Note: For the bash runs, I have excluded the `strace` output of bash starting and running Hadolint.

based on this, all that is needed is to find what happens between `mmap(0x400000, 19647156, PROT_READ|PROT_WRITE|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x400000` and `mprotect(0x400000, 19647156, PROT_READ|PROT_EXEC) = 0` since that is where Hadolint crashes.

I tried to disassemble the binary to see what was going on, but since it is a Haskell program the binary is just the runtime and object code for Haskell, not the assembly that can be reversed. I'm not very familiar with Haskell, but if there is a way to produce a pure assembly output for Hadolint (something that can produce an output with `objdump -D ./hadolint` that would help significantly with the debugging.

**Comment by lorenzo** [comment]
@tanishq-dubey try building `hadolint` locally and see if it makes any difference. We are compressing the binary with `upx` so that could also be a possible source of the problem.

So far I've been unable to replicate that problem, running the latest version compiled on my Mac.

**Comment by lorenzo** [comment]
@omh1280 I released a new version with support for parallel linters, in case it helps with your issue

**Comment by m-ildefons** [comment]
@lorenzo , could you please elaborate on how the binaries are packed with `upx`? Their issue tracker has an issue of similar type: https://github.com/upx/upx/issues/385
And when I compress a local build with: `upx --brute`, I too get exit codes 127 even without docker.
Meanwhile uncompressing hadolint manually gets rid of the error as far as I can tell.

```Dockerfile
FROM ubuntu:focal

RUN apt-get update && apt-get install --no-install-recommends -y \
  wget=1.20.3-1ubuntu1 \
  ca-certificates=20190110ubuntu1.1 \
  parallel=20161222-1.1 \
  tar=1.30+dfsg-7 \
  xz-utils=5.2.4-1ubuntu1 \
  && rm -rf /var/lib/apt/lists
RUN wget -q https://github.com/hadolint/hadolint/releases/download/v1.18.0/hadolint-Linux-x86_64 -O hadolint \
  && chmod +x hadolint
RUN wget -q https://github.com/upx/upx/releases/download/v3.96/upx-3.96-amd64_linux.tar.xz \
  && xz -d upx-3.96-amd64_linux.tar.xz && tar -xf upx-3.96-amd64_linux.tar \
  && mv upx-3.96-amd64_linux/upx /upx
RUN /upx -d /hadolint -ononupxhadolint

COPY ./Dockerfile ./Dockerfile

# hadolint ignore=SC2039
CMD [ "/bin/bash", "-c", "parallel -N0 --halt now,fail=1 /nonupxhadolint /Dockerfile ::: {1..1000}; echo $?" ]
```

This also has the nice benefit that the uncompressed hadolint runs a **lot** faster

**Comment by zemanlx** [comment]
@m-ildefons See https://github.com/hadolint/hadolint/blob/master/.travis.yml#L33

**Comment by tanishq-dubey** [comment]
I can run my testbench with the locally built binary in a few hours. Will post the results.

**Comment by omh1280** [comment]
> This also has the nice benefit that the uncompressed hadolint runs a lot faster

Sounds like upx is the cause of the bug and makes hadolint run slower? Is that right?

**Comment by zemanlx** [comment]
@omh1280 IMHO for most people speed does not really matter as they are testing one or a few Dockerfiles and downloading binary in CI usually takes longer than running it, which why we prefer having smaller binary compressed by `upx`. 

Therefore I would prefer not to ditch `upx`, but maybe changing option from `--brutal` to `--best` as it is suggested in https://github.com/upx/upx/issues/385.

**Comment by tanishq-dubey** [comment]
Hey all,

Sorry for delivering late on results, but I reran my tests with a non compressed hadolint. 

Using the same test benches (just changing `hadolint-Linux-x86_64` to my executeable based on commit `bf248d3b93e9741111488f8c81191b38d20cba46`) I was unable to reproduce the `127` exit code across a total of 1330802 runs. In addition, 5000 `strace`'ed runs did not produce a `127` error code. 

I can run the testbench with the `--best` compression as well. Just need instructions on how to compress.

Best,

**Comment by lorenzo** [comment]
@tanishq-dubey did you find out if using `--best` had the same effects?

**Comment by dogmatic69** [comment]
Can this be reopened? We experience occasional exits of 127 using github actions with the latest hadolint/hadolint docker image.

**Comment by lorenzo** [comment]
sure, can you please help determine if that’s the case with a previous version?

**Comment by dogmatic69** [comment]
Hi :)

Give me a version to try and I'll run it for a while

**Comment by dogmatic69** [comment]
bump
